consensus/hotstuff/signature.go

```diff
@@ -8,37 +8,53 @@ import (
+	// TrustedAdd adds the signature share to the reconstructors internal
+	// state. Validity of signature is not checked. It is up to the
+	// implementation, whether it still adds a signature or not, when the
+	// minimal number of required sig shares has already been reached,
+	// because the reconstructed group signature is the same.
+	// Returns: true if and only if enough signature shares were collected
 	TrustedAdd(signerID flow.Identifier, sig crypto.Signature) (bool, error)
+
+	// HasSufficientShares returns true if and only if reconstructor
+	// has collected a sufficient number of signature shares.
 	HasSufficientShares() bool
-	// Reconstruct the beacon signature if there are enough shares. It errors if not enough shares were collected.
+
+	// Reconstruct reconstructs the group signature from the provided
+	// signature shares. Errors if the the number of shares is insufficient
+	// or some of the added signatures shares were invalid.
 	Reconstruct() (crypto.Signature, error)
 }
 
-// RandomBeaconAggregator aggregates the random beacon signatures
-type RandomBeaconAggregator interface {
-	// TrustedAdd adds an already verified signature.
-	// return (true, nil) means the signature has been added
-	// return (false, nil) means the signature is a duplication
-	TrustedAdd(signerID flow.Identifier, sig crypto.Signature) (bool, error)
-	// Aggregate assumes enough shares have been collected, it aggregates the signatures
-	// and return the aggregated signature.
-	// if called concurrently, only one thread will be running the aggregation.
-	Aggregate() ([]byte, error)
-}
+// SigType is the aggregable signature type.
+type SigType int
+
+// There are two signatures are aggregable, one is the normal staking signature,
+// the other is the threshold sig used as staking sigature.
+const (
+	SigTypeStaking SigType = iota
+	SigTypeRandomBeacon
+	SigTypeInvalid
+)
 
 // SignatureAggregator aggregates the verified signatures.
 // It can either be used to aggregate staking signatures or aggregate random beacon signatures,
 // but not a mix of staking signature and random beacon signature.
+// Implementation of SignatureAggregator must be concurrent safe
 type SignatureAggregator interface {
-	// TrustedAdd adds an already verified signature.
-	// return (true, nil) means the signature has been added
-	// return (false, nil) means the signature is a duplication
-	TrustedAdd(signerID flow.Identifier, sig crypto.Signature) (bool, error)
+	// TrustedAdd adds an already verified signature, and look up the weight for the given signer,
+	// and add it to the total weight, and returns the total weight that have been collected.
+	// return (true, 1000, nil) means the signature has been added, and 1000 weight has been collected in total.
+	// return (false, 1000, nil) means the signature is a duplication and 1000 weight has been collected in total.
+	TrustedAdd(signerID flow.Identifier, sig crypto.Signature) (added bool, totalWeight uint64, exception error)
```

Comment:
I would omit `added bool` unless you know a case where it will be useful

Reply:
I wonder how are we going to handle duplicated votes. 

Is it possible to skip some logic for duplicated votes? If so, this field might be useful.

Reply:
We can do this on `VoteAggregator` level or `VoteCollector`, don't really have a preference.

Reply:
OK, I can remove it for now.